hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -271,29 +273,32 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       try {
         persistTable(tbl, updateHiveTable);
         commitStatus = CommitStatus.SUCCESS;
-      } catch (Throwable persistFailure) {
+      } catch (org.apache.hadoop.hive.metastore.api.AlreadyExistsException e) {
+        throw new AlreadyExistsException(e, "Table already exists: %s.%s", database, tableName);
+
+      } catch (InvalidObjectException e) {
+        throw new ValidationException(e, "Invalid table name for %s.%s", database, tableName);
```

Comment:
This scares me a bit more, is InvalidObjectException only thrown for this sort of naming issue?

Reply:
Yea so I grepped through the instances it throws in Hive code base, it is always like Precondition check in Iceberg for user-error, and never proceeds further.

This is a github search (which unfortunately returns only a few top results):

https://github.com/search?q=%22throw+new+InvalidObjectException*%22+repo%3Aapache%2Fhive&type=code

My grep finds it thrown only in these classes which you can also double-check inside:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/SessionHiveMetaStoreClient.java
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/StatObjectConverter.java
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java


It should not be thrown when Hive commits successfully, or disconnect issue, if that is the fear?  Maybe @pvary can verify?

Reply:
It's the unknown unknowns I worry about. I just don't have enough familiarity with the codebase to even know where I should be looking or If I can guarantee behaviors in the future.

Reply:
So I'll try to explain the HMS flow (using the latest Hive release tag github links to make it persistent):

Every call goes through HMSHandler.  For these two calls (CreateTable or AlterTable), it goes like:

[HMSHandler::create_table_core](https://github.com/apache/hive/blob/8190d2be7b7165effa62bd21b7d60ef81fb0e4af/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L1789) -> [ObjectStore::createTable](https://github.com/apache/hive/blob/8190d2be7b7165effa62bd21b7d60ef81fb0e4af/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L1298), the call being at this line : ms.createTable()

[HMSHandler::alter_table_core](https://github.com/apache/hive/blob/8190d2be7b7165effa62bd21b7d60ef81fb0e4af/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L4998) -> [ObjectStore::alterTable](https://github.com/apache/hive/blob/8190d2be7b7165effa62bd21b7d60ef81fb0e4af/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L4005) , the call being at the line ms.alterTable()

After this it uses JDO persistence library which does not throw any InvalidObjectException.

So we check the four places.  InvalidObjectException is only ever thrown in the first one (HMSHandler::create_table_core).  

Note that ObjectStore methods do call convertToMTable() that does throw the InvalidObjectException if database is null, but in Iceberg case this is actually checked way before it gets here (can share the stack trace if you want).

Not sure if this helps?

Reply:
The alternative, if we are very paranoid, can check the exact exception message to make sure its the exception we want.  Would be a bit hacky but it should be the same purpose as this block is to expect only that exception.

Reply:
Makes sense, maybe @pvary could take a look ?

Reply:
I do share Russell's concern as `InvalidObjectException` sounds like something that can be returned not only when the table is not there. However, I also don't know the Hive codebase. I'd ask @pvary.

@szehon-ho, how did you hit this issue, btw? Do you have an example?

Reply:
Hive error handling is not something I would be proud of. There are plenty of inconsistencies around the code, and the last time I have tried to clean this up, I have faced serious resistance. The reasoning behind it was that the users of HMS are plenty and they might build above this haphazard error handling.

That said:
- InvalidObjectException is generally used for any error in the provided objects:
  - `You must specify a path for the catalog`
  - `BLABLA is not a valid catalog name`
  - `Invalid type name`
  - `Invalid column ...`
- The error could signal missing objects as well - typically when storing related objects:
  - `No such catalog BLABLA`
  - `Unable to add partition because table or database do not exist`
  - `Database BLABLA doesn't exist`
  
So I would definitely opt for a more general error message, maybe mentioning that the most common error is the wrong table name. All that said, I think the `ValidationException` seems ok, since there is some validation error happened when we tried to store the table.

Reply:
Thanks guys for the comment, yea I will definitely change the message.  @aokolnychyi does the description in the issue explain it?

Reply:
Sounds like we have a consensus that it is ok to catch and rethrow InvalidObjectException these and not change them to commit state unknown. The only thing that remains is the error message.

Reply:
Done